### PR TITLE
Improvements: Add missing Flink TypeInformation for proper serialization

### PR DESCRIPTION
**Pull request by lciolecki**

### CODE REVIEW

File: docs/MigrationGuide.md [hunk 1]
```diff
@@ -51,6 +51,10 @@ To see the biggest differences please consult the [changelog](Changelog.md).
 * [#6952](https://github.com/TouK/nussknacker/pull/6952) Improvement: TypeInformation support for scala.Option:
   If you used CaseClassTypeInfoFactory with case classes that contain the Option type, the state won't be restored after the upgrade.
 
+* [#7058](https://github.com/TouK/nussknacker/pull/7058) Improvements: Add missing Flink TypeInformation for proper serialization:
+  * In case of using base (bounded and unbounded) Flink components state will be probably not compatible
+  * `FlinkCustomNodeContext.typeInformationDetection` hase been removed, please use `TypeInformationDetection.instance` instead
```
Comment (lciolecki) [inline docs/MigrationGuide.md:56]: Typo..

File: engine/flink/executor/src/main/scala/pl/touk/nussknacker/engine/process/registrar/PartReferenceTypeInformation.scala [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    s"PartReferenceTypeInformation"
+
+  override def canEqual(obj: Any): Boolean =
+    obj.isInstanceOf[PartReferenceTypeInformation]
+
+  override def equals(obj: Any): Boolean =
+    canEqual(obj)
+
+  override def hashCode(): Int =
+    Objects.hashCode(this)
```
Comment (lciolecki) [inline engine/flink/executor/src/main/scala/pl/touk/nussknacker/engine/process/registrar/PartReferenceTypeInformation.scala:43]: Out this is stupid ;)

File: engine/flink/schemed-kafka-components-utils/src/main/scala/pl/touk/nussknacker/engine/schemedkafka/sink/flink/FlinkKafkaUniversalSink.scala [hunk 3]
```diff
@@ -41,6 +41,7 @@ class FlinkKafkaUniversalSink(
       dataStream: DataStream[ValueWithContext[Value]],
       flinkNodeContext: FlinkCustomNodeContext
   ): DataStreamSink[_] =
+    // FIXME: Missing map TypeInformation
```
Comment (arkadius) [inline engine/flink/schemed-kafka-components-utils/src/main/scala/pl/touk/nussknacker/engine/schemedkafka/sink/flink/FlinkKafkaUniversalSink.scala:44]: You want to fix this in this PR or it is a TODO for someone else?

**Review by arkadius — COMMENTED** [review]
Nice change, some minor comments added

File: engine/flink/schemed-kafka-components-utils/src/main/scala/pl/touk/nussknacker/engine/schemedkafka/sink/flink/FlinkKafkaUniversalSink.scala (hunk 3 above)
Comment (lciolecki) [reply engine/flink/schemed-kafka-components-utils/src/main/scala/pl/touk/nussknacker/engine/schemedkafka/sink/flink/FlinkKafkaUniversalSink.scala:44]: I'll do it later -> next PR.

File: engine/flink/executor/src/main/scala/pl/touk/nussknacker/engine/process/registrar/PartReferenceTypeInformation.scala [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        target.writeUTF(r.branchId)
+      case r: DeadEndReference =>
+        target.writeShort(DeadEndReferenceType)
+        target.writeUTF(r.nodeId)
+      case r: EndReference =>
+        target.writeShort(EndReferenceType)
+        target.writeUTF(r.nodeId)
+      case r: FragmentEndReference =>
+        target.writeShort(FragmentEndReferenceType)
+        // TODO: Maybe we can use here TypeInformationDetection.instance.forType(Typed.fromInstance(r)).createSerializer ??
```
[the remark this answers is not in the document]
Comment (lciolecki) [reply engine/flink/executor/src/main/scala/pl/touk/nussknacker/engine/process/registrar/PartReferenceTypeInformation.scala:108]: Ooou, grat idea! I'll try to pass to FragmentEndReference TypeResult.

File: docs/Changelog.md [hunk 5]
```diff
@@ -76,6 +76,7 @@
 * [#6958](https://github.com/TouK/nussknacker/pull/6958) Add message size limit in the "Kafka" exceptionHandler
 * [#6988](https://github.com/TouK/nussknacker/pull/6988) Remove unused API classes: `MultiMap`, `TimestampedEvictableStateFunction`
 * [#7000](https://github.com/TouK/nussknacker/pull/7000) Show all possible options for dictionary editor on open.
+* [#7058](https://github.com/TouK/nussknacker/pull/7058) Improvements: Add missing Flink TypeInformation for proper serialization
```
Comment (arkadius) [inline docs/Changelog.md:79]: Maybe: `Performance optimization: ..."? For user this is the most important

File: engine/flink/executor/src/main/scala/pl/touk/nussknacker/engine/process/registrar/PartReferenceTypeInformation.scala [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+  private val NextPartReferenceType    = 1
+  private val JoinReferenceType        = 2
+  private val DeadEndReferenceType     = 3
+  private val EndReferenceType         = 4
+  private val FragmentEndReferenceType = 5
+}
+
+// TODO: It's a experimental version, there can be some issues with this serializer
+// TODO: For FragmentEndReference we still use KryoSerializer :(
+class PartReferenceSerializer(val fragmentEndReferenceSerializer: KryoSerializer[FragmentEndReference])
```
[the remark this answers is not in the document]
Comment (lciolecki) [reply engine/flink/executor/src/main/scala/pl/touk/nussknacker/engine/process/registrar/PartReferenceTypeInformation.scala:55]: Can we do it later, in other PR? :)

File: engine/flink/executor/src/main/scala/pl/touk/nussknacker/engine/process/registrar/PartReferenceTypeInformation.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (lciolecki) [reply engine/flink/executor/src/main/scala/pl/touk/nussknacker/engine/process/registrar/PartReferenceTypeInformation.scala:108]: The same as above

File: engine/flink/executor/src/main/scala/pl/touk/nussknacker/engine/process/registrar/PartReferenceTypeInformation.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (arkadius) [reply engine/flink/executor/src/main/scala/pl/touk/nussknacker/engine/process/registrar/PartReferenceTypeInformation.scala:55]: But it is a small thing. Maybe we misunderstand each other?

**Review by arkadius — APPROVED** [review]

**Pull request merged**